lib/srv/git/github.go

```diff
@@ -30,35 +34,141 @@ import (
 	actualFingerprint := ssh.FingerprintSHA256(key)
-	if slices.Contains(knownGithubDotComFingerprints, actualFingerprint) {
-		return nil
+	for _, fingerprint := range g.getKnownFingerprints() {
+		if sshutils.EqualFingerprints(actualFingerprint, fingerprint) {
+			return nil
+		}
 	}
 	return trace.BadParameter("cannot verify github.com: unknown fingerprint %v algo %v", actualFingerprint, key.Type())
 }
 
+func (g *githubFingerprintManager) getKnownFingerprints() []string {
+	const refreshDuration = time.Hour
+	g.mu.RLock()
+	if g.clock.Now().Sub(g.lastCheck) < refreshDuration {
+		defer g.mu.RUnlock()
+		return g.fingerprints
+	}
+	g.mu.RUnlock()
+
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	if g.clock.Now().Sub(g.lastCheck) < refreshDuration {
+		return g.fingerprints
+	}
+
+	logger := slog.With(teleport.ComponentKey, "git:github")
+
+	// Check if eTag is the same to avoid downloading the whole thing which
+	// contains a lot of irrelevant info.
+	if g.etag != "" {
+		etag, err := g.client.fetchETag()
+		switch {
+		case err != nil:
+			logger.WarnContext(context.Background(), "Failed to fetch eTag from GitHub meta API", "error", err)
+			// Don't give up yet if HEAD fails.
+
+		case etag == g.etag:
+			g.lastCheck = g.clock.Now()
+			logger.DebugContext(context.Background(), "ETag did not change for GitHub meta API")
+			return g.fingerprints
+
+		default:
+			logger.DebugContext(context.Background(), "ETag changed for GitHub meta API", "new", etag)
+		}
+	}
+
+	fingerprints, etag, err := g.client.fetchFingerprints()
+	if err != nil {
+		logger.WarnContext(context.Background(), "Failed to fetch fingerprints from GitHub meta API", "error", err)
+		return g.fingerprints
+	}
+	logger.DebugContext(context.Background(), "Found SSH fingerprints from Github meta API", "fingerprints", fingerprints, "etag", etag)
+	g.etag = etag
+	g.fingerprints = fingerprints
+	g.lastCheck = g.clock.Now()
+	return g.fingerprints
+}
+
+var githubFingerprints = newGithubFingerprintManager()
+
+type githubMetadataHTTPClient struct {
+	api    string
+	client *http.Client
+}
+
+func newGithubMetadataTTPClient() *githubMetadataHTTPClient {
+	return &githubMetadataHTTPClient{
+		api: "https://api.github.com/meta",
+		client: &http.Client{
+			Timeout: defaults.HTTPRequestTimeout,
+		},
+	}
+}
+
+func (c *githubMetadataHTTPClient) fetchETag() (string, error) {
+	resp, err := http.Head(c.api)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	return resp.Header.Get("ETag"), nil
+}
+
+func (c *githubMetadataHTTPClient) fetchFingerprints() ([]string, string, error) {
+	resp, err := http.Get(c.api)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+	defer resp.Body.Close()
+
+	// Meta API reference:
+	// https://docs.github.com/en/rest/meta/meta?apiVersion=2022-11-28#get-github-meta-information
+	meta := struct {
+		// Fingerprints lists the fingerprints by algo type.
+		Fingerprints map[string]string `json:"ssh_key_fingerprints"`
```

Comment:
I would rather just compare against the actual host keys parsed from `ssh_keys` - they're not much bigger to handle than the fingerprints, after all, especially if we keep the fingerprints as hex or base64.